config/v1/types_oauth.go

```diff
@@ -1,37 +1,518 @@
+
+	// Google enables user authentication using Google credentials
+	// +optional
+	Google *GoogleIdentityProvider `json:"google,omitempty"`
+
+	// OpenID enables user authentication using OpenID credentials
+	// +optional
+	OpenID *OpenIDIdentityProvider `json:"openID,omitempty"`
+}
+
+// BasicAuthPasswordIdentityProvider provides identities for users authenticating using HTTP basic auth credentials
+type BasicAuthPasswordIdentityProvider struct {
+	// RemoteConnectionInfo contains information about how to connect to the external basic auth server
+	OAuthRemoteConnectionInfo `json:",inline"`
+}
+
+// RemoteConnectionInfo holds information necessary for establishing a remote connection
+type OAuthRemoteConnectionInfo struct {
+	// URL is the remote URL to connect to
+	URL string `json:"url"`
+	// CA is a reference to a ConfigMap containing the CA for verifying TLS connections
+	CA ConfigMapReference `json:"ca"`
+
+	// TLSClientCert references a secret containing the TLS client certificate to present when
+	// connecting to the server.
+	// Looks under the key "tls.cert" for the data unless a lookup key is specified in the secret ref
+	TLSClientCert LocalSecretReference `json:"tlsClientCert"`
+
+	// TLSClientKey references a secret containing the TLS private key for the client certificate
+	// Looks under the key "tls.key" for the data unless a lookup key is specified in the secret ref
+	TLSClientKey LocalSecretReference `json:"tlsClientKey"`
+}
+
+// HTPasswdDataKey is the default key for the htpasswd file data in a secret
+const HTPasswdDataKey = "htpasswd"
+
+// HTPasswdPasswordIdentityProvider provides identities for users authenticating using htpasswd credentials
+type HTPasswdPasswordIdentityProvider struct {
+	// Data is a reference to a secret containing the data to use as the htpasswd file
+	// Looks under the key `htpasswd` unless a lookup key is specified in the secret ref
+	FileData LocalSecretReference `json:"fileData"`
+}
+
+const (
+	// BindPasswordKey is default the key for the LDAP bind password in a secret
+	BindPasswordKey = "bindPassword"
+	// ClientSecretKey is the key for the oauth client secret data in a secret
+	ClientSecretKey = "clientSecret"
+)
+
+// LDAPPasswordIdentityProvider provides identities for users authenticating using LDAP credentials
+type LDAPPasswordIdentityProvider struct {
+	// URL is an RFC 2255 URL which specifies the LDAP search parameters to use.
+	// The syntax of the URL is:
+	//    ldap://host:port/basedn?attribute?scope?filter
+	URL string `json:"url"`
 
+	// BindDN is an optional DN to bind with during the search phase.
+	// +optional
+	BindDN string `json:"bindDN"`
+
+	// BindPasswordSecretRef is a reference to the secret containing an optional password to bind
+	// with during the search phase.
+	// Looks under the key `bindPassword` unless a lookup key is specified in the secret ref
+	// +optional
+	BindPasswordSecretRef LocalSecretReference `json:"bindPasswordSecretRef"`
```

Comment:
followup: We should name fields for the user intent, not how they fulfill it.  just `bindPassword`

Reply:
Reason I did `SecretRef` was that if you have:
```yaml
bindPassword:
  name: bind-password
```
it's not obvious that's the name of a secret.

Reply:
> it's not obvious that's the name of a secret.

The openapi and `oc explain` list the type.  We don't follow hungarian notation.